buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -217,24 +246,36 @@ public CompositeByteBuf addComponents(int cIndex, ByteBuf... buffers) {
     }
 
     private int addComponents0(int cIndex, ByteBuf... buffers) {
-        checkComponentIndex(cIndex);
-
-        if (buffers == null) {
-            throw new NullPointerException("buffers");
-        }
+        checkNotNull(buffers, "buffers");
+        int i = 0;
+        try {
+            checkComponentIndex(cIndex);
 
-        // No need for consolidation
-        for (ByteBuf b: buffers) {
-            if (b == null) {
-                break;
+            // No need for consolidation
+            for (; i < buffers.length; ++i) {
+                ByteBuf b = buffers[i];
+                if (b == null) {
+                    break;
+                }
+                cIndex = addComponent0(cIndex, b) + 1;
+                int size = components.size();
+                if (cIndex > size) {
+                    cIndex = size;
+                }
             }
-            cIndex = addComponent0(cIndex, b) + 1;
-            int size = components.size();
-            if (cIndex > size) {
-                cIndex = size;
+            return cIndex;
+        } finally {
+            for (; i < buffers.length; ++i) {
+                ByteBuf b = buffers[i];
+                if (b != null) {
+                    try {
+                        b.release();
```

Comment:
I originally had this, but figured we could avoid the instanceof check and casting.